pkg/referrerstore/oras/authprovider/azure/azure_helper.go

```diff
@@ -0,0 +1,266 @@
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+	"unicode"
+
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+)
+
+const (
+	userAgentHeader = "User-Agent"
+	userAgent       = "kubernetes-credentialprovider-acr"
+	maxReadLength   = 10 * 1 << 20 // 10MB
+)
+
+var client = &http.Client{
+	Transport: utilnet.SetTransportDefaults(&http.Transport{}),
+	Timeout:   time.Second * 10,
+}
+
+type authDirective struct {
+	service string
+	realm   string
+}
+
+type acrAuthResponse struct {
+	RefreshToken string `json:"refresh_token"`
+}
+
+// formats and sends a challenge request to the server address
+func receiveChallengeFromLoginServer(serverAddress, scheme string) (*authDirective, error) {
+	challengeURL := url.URL{
+		Scheme: scheme,
+		Host:   serverAddress,
+		Path:   "v2/",
+	}
+	var err error
+	var r *http.Request
+	r, err = http.NewRequest("GET", challengeURL.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct request, got %v", err)
+	}
+	r.Header.Add(userAgentHeader, userAgent)
+
+	var challenge *http.Response
+	if challenge, err = client.Do(r); err != nil {
+		return nil, fmt.Errorf("error reaching registry endpoint %s, error: %s", challengeURL.String(), err)
+	}
+	defer challenge.Body.Close()
+
+	if challenge.StatusCode != 401 {
+		return nil, fmt.Errorf("registry did not issue a valid AAD challenge, status: %d", challenge.StatusCode)
+	}
+
+	var authHeader []string
+	var ok bool
+	if authHeader, ok = challenge.Header["Www-Authenticate"]; !ok {
+		return nil, fmt.Errorf("challenge response does not contain header 'Www-Authenticate'")
+	}
+
+	if len(authHeader) != 1 {
+		return nil, fmt.Errorf("registry did not issue a valid AAD challenge, authenticate header [%s]",
+			strings.Join(authHeader, ", "))
+	}
+
+	authSections := strings.SplitN(authHeader[0], " ", 2)
+	authType := strings.ToLower(authSections[0])
+	var authParams *map[string]string
+	if authParams, err = parseAssignments(authSections[1]); err != nil {
+		return nil, fmt.Errorf("unable to understand the contents of Www-Authenticate header %s", authSections[1])
+	}
+
+	// verify headers
+	if !strings.EqualFold("Bearer", authType) {
+		return nil, fmt.Errorf("Www-Authenticate: expected realm: Bearer, actual: %s", authType)
+	}
+	if len((*authParams)["service"]) == 0 {
+		return nil, fmt.Errorf("Www-Authenticate: missing header \"service\"")
+	}
+	if len((*authParams)["realm"]) == 0 {
+		return nil, fmt.Errorf("Www-Authenticate: missing header \"realm\"")
+	}
+
+	return &authDirective{
+		service: (*authParams)["service"],
+		realm:   (*authParams)["realm"],
+	}, nil
+}
+
+// exchanges the specifed access token for a registry refresh token
+func performTokenExchange(
+	serverAddress string,
+	directive *authDirective,
+	tenant string,
+	accessToken string) (string, error) {
+	var err error
+	data := url.Values{
+		"service":       []string{directive.service},
+		"grant_type":    []string{"access_token_refresh_token"},
```

Comment:
Why is the access token passed as refresh token? can the grant type be access token?